extension/intents/nicknames/nicknames.js

```diff
@@ -114,3 +102,89 @@ intentRunner.registerIntent({
     intentRunner.unregisterPageName(name);
   },
 });
+
+intentRunner.registerIntent({
+  name: "nicknames.pause",
+  async run(context) {
+    if (context.routineExecutor === undefined) {
+      const exc = new Error("Command not available.");
+      exc.displayMessage = "Command not available.";
+      throw exc;
+    }
+    browser.runtime.sendMessage({
+      type: "presentMessage",
+      message: context.slots.message,
+    });
+    context.routineExecutor.pauseRoutine();
+  },
+});
+
+intentRunner.registerIntent({
+  name: "nicknames.continue",
+  async run() {
+    const { pausedRoutine } = await browser.storage.local.get("pausedRoutine");
+    if (pausedRoutine === undefined) {
+      const exc = new Error(
+        "'Continue' can only be used in a routine after 'pause routine'"
+      );
+      exc.displayMessage =
+        "'Continue' can only be used in a routine after 'pause routine'";
+      throw exc;
+    }
+
+    const { name, nextIndex, states, forIndex } = pausedRoutine;
+    const registeredNicknames = await intentRunner.getRegisteredNicknames();
+    const routineExecutor = new RoutineExecutor(
+      registeredNicknames[name].nickname,
+      registeredNicknames[name].contexts,
+      nextIndex,
+      states,
+      forIndex
+    );
+    await browser.storage.local.remove("pausedRoutine");
+    await routineExecutor.run();
+  },
+});
+
+intentRunner.registerIntent({
+  name: "nicknames.startForLoop",
+  async run(context) {
+    const objectToLoop = [];
+    if (context.parameters.option === "clipboard") {
+      const result = await navigator.clipboard.readText();
+      const splitResult = result.split("\n");
+      for (const key of splitResult) {
+        objectToLoop.push({ value: key });
+      }
+    }
+
+    if (context.parameters.option === "selected") {
+      const selectedTabs = await browser.tabs.query({ highlighted: true });
+      for (const tab of selectedTabs) {
+        objectToLoop.push({ selected: tab.url });
```

Comment:
Using strings here for now is probably fine, but ultimately there should be special wrappers around each type. For instance, you might want to do:

```
for each selected tab x
activate tab x
pause
close tab x
end for
```

For some ways of using the tab the URL is appropriate, for others the tab object or tab ID is appropriate.

I think it's best _not_ to use the actual raw tab object, because then consumers have to test a bunch of types to figure things out. But instead it would be a Tab object with a `.toString() {return this._tab.url}` method, and maybe some other methods specific to tabs, or it inherits from an abstract base class that implements tab-related fallbacks for non-tab objects.

But this can be a followup.

Reply:
An example of how to create a tab object that  an be used specially for some intents, but treated like a URL by default:

```js
class TabVariable {
  constructor(tabObj) {
    this.tabObj = tabObj;
  }
  toString() {
    return this.tabObj.url;
  }
  closeTab() {
    return browser.tabs.remove(this.tabObj.id);
  }
}

registerIntent("tabs.closeTab", async (context) => {
  if (context.slots.query.closeTab) {
    await context.slots.query.closeTab();
  }
});
```